ssg/utils.py

```diff
@@ -333,22 +333,30 @@ def enum(*args):
     return type('Enum', (), enums)
 
 
-def apply_formatting_on_dict_values(source_dict, string_dict, ignored_keys=frozenset()):
+def recurse_or_substitute_or_do_nothing(
+        v, string_dict, ignored_keys=frozenset()):
+    if isinstance(v, dict):
+        return apply_formatting_on_dict_values(v, string_dict, ignored_keys)
+    elif isinstance(v, str):
+        return v.format(**string_dict)
+    else:
+        return v
+
```

Comment:
@evgenyz Where the frozenset should be after the refactoring to make it work as expected? Should it be in `recurse_or_substitute_or_do_nothing` or in the `apply_formatting_on_dict_values`?

Reply:
Preferably in both places, so the default argument to `ignored_keys` is an empty iterable that is not mutable. That way, you even won't have to set `ignored_keys` to that empty list.
However, the decomposition into those two functions in this form was not such great idea, because the `recurse_or_substitute_or_do_nothing` accepts `ignored_keys` that it doesn't care about - it just passes those on, and it also hardcodes "recursion" to calling `apply_formatting_on_dict_values`, which creates a circular dependency. So I would say that having seen this, I would either revert to the original code, or if you became sold to the idea of decomposing handling of container and handling of items, I would pass `recurse_or_substitute_or_do_nothing` a callback, e.g. `lambda value, substitutions: recurse_or_substitute_or_do_nothing(value, substitutions, ignored_keys)` that would be used in the item-processing function if the item passed is a dictionary.
So sorry, suggesting the decomposition like this was a mistake.